src/glib/mine/svm.h

```diff
@@ -13,376 +13,640 @@
-
-    svm_problem_t svm_problem;
-    svm_problem.l = AllN;
-    svm_problem.y = (double *)malloc(AllN*sizeof(double));
-
-    svm_problem.x = (svm_node_t **)malloc(AllN*sizeof(svm_node_t *));
-    svm_node_t* x_space = (svm_node_t *)malloc((AllN*(DimN+1))*sizeof(svm_node_t));
-    int N = 0, prevN = 0;
-    for (int Idx = 0; Idx < AllN; ++Idx) { // # of examples
-        prevN = N;
-        svm_problem.y[Idx] = TargetV[Idx];
-        for (int Jdx = 0; Jdx < DimN; ++Jdx) { // # of features
-            if (VecV.At(Jdx, Idx) != 0.0) {
-                x_space[N].index = Jdx+1;
-                x_space[N].value = VecV.At(Jdx, Idx);
-                ++N;
+        // train the model
+        svm_model_t* svm_model = svm_train(&svm_problem, &svm_parameter, DebugNotify, ErrorNotify);
+        
+        TFltVV SVs(svm_model->l, Dim);
+        TFltVV Coef(svm_model->nr_class - 1, svm_model->l);
+        TFltV Rho(svm_model->nr_class * (svm_model->nr_class - 1)/2);
+        TFltV ProbA(svm_model->nr_class * (svm_model->nr_class - 1)/2);
+        TFltV ProbB(svm_model->nr_class * (svm_model->nr_class - 1)/2);
+        TIntV SVIndices(svm_model->l);
+        TIntV Label(svm_model->nr_class);
+        TIntV nSV(svm_model->nr_class);
+
+        // compute normal vector from support vectors
+        TFltV WgtV(Dim);
+        TFlt Bias = -svm_model->rho[0]; // LIBSVM does w*x-b, while we do w*x+b; thus the sign flip
+        EAssertR(TLinAlg::Norm(WgtV) == 0.0, "Expected a zero weight vector.");
+        for (int Idx = 0; Idx < svm_model->l; ++Idx) {
+            svm_node_t* SV = svm_model->SV[Idx];
+            while (SV->index != -1) {
+                SVs.PutXY(Idx, SV->index - 1, SV->value);
+                WgtV[SV->index - 1] += svm_model->sv_coef[0][Idx] * SV->value;
+                ++SV;
+            }
+            SVIndices.SetVal(Idx, svm_model->sv_indices[Idx]);
+            for (int cIdx = 0; cIdx < svm_model->nr_class - 1; cIdx++){
+              Coef.PutXY(cIdx, Idx, svm_model->sv_coef[cIdx][Idx]);
             }
+        }        
+        for (int Idx = 0; Idx < svm_model->nr_class * (svm_model->nr_class - 1)/2; Idx++){
+            Rho.SetVal(Idx, svm_model->rho[Idx]);
+        }          
+        for (int Idx = 0; Idx < svm_model->nr_class; Idx++){
+          Label.SetVal(Idx, svm_model->label[Idx]);
+          nSV.SetVal(Idx, svm_model->nSV[Idx]);
         }
-        x_space[N].index = -1;
-        ++N;
-        svm_problem.x[Idx] = &x_space[prevN];
+        
+        this->WgtV = WgtV;
+        this->Bias = Bias;
+        this->NClass = svm_model->nr_class;
+        this->NSV = svm_model->l;
+        this->SV = SVs;
+        this->Coef = Coef;
+        this->Rho = Rho;
+        this->ProbA = ProbA;
+        this->ProbB = ProbB;
+        this->SVIndices = SVIndices;
+        this->Label = Label;
+        this->nSV = nSV;
+        this->Free = svm_model->free_sv;
+
+        svm_free_and_destroy_model(&svm_model);
```

Comment:
I guess that answers my question above :-) Please add comment whenever you call malloc with when and where it will be freed or who is responsible for freeing the memory.